Product Collection: Add "order" reference to Upsells and Cross-sells

imanish003 opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #52019

This PR adds support for "order" reference for collections. In this PR, I have added "order" reference for "Upsells" and "Cross-Sells" collections, but this can be extended for other collections as well. 

When order context is available, "Upsells" and "Cross-Sells" collections will show UI in Inspector control with radio button named "From products in the order". This way, merchants can choose to show upsells/cross-sells based on the order. Here is how it looks:
![image](https://github.com/user-attachments/assets/577a2134-1c76-44b7-b204-871a0e545f32)



<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

**Case 1: Checkout with One Product**
1. Go to "Order Confirmation" template
2. Add Product Collection block and select "Upsells" collection
3. In Editor, verify that:
	- "From products in the order" is selected by default in Inspector controls as shown in screenshot below.
	- Preview label is showing
![Image](https://github.com/user-attachments/assets/d4bea63d-64c3-4b5c-8bb6-1e6806b0cf63)

4. Save template
5. Go to frontend and add one product to cart which has upsells products.
6. Checkout and you will be redirected to "Order Confirmation" page.
7. On "Order Confirmation" page, verify that "Upsells" collection is showing upsells products from the order. In other words, Upsells products will be shown for products which are there in the order.
8. Perform same testing for "Cross-Sells" collection.

**Case 2: Checkout with Multiple Products**
1. 1-4 steps are the same as in Case 1.
2. Go to frontend and add multiple products to cart. Each of these products should have upsells products.
3. Checkout and you will be redirected to "Order Confirmation" page.
4. On "Order Confirmation" page, verify that "Upsells" collection is showing merged upsells products from all the products in the order. For example, if product A has 2 upsells products and product B has 3 upsells products, then "Upsells" collection should have 5 upsells products in total. Also, There shouldn't be any duplicates in the "Upsells" collection.
5. Perform same testing for "Cross-Sells" collection.

**Case 3: From a specific product**
1. On "Order Confirmation" template, add Product Collection block and select "Upsells" collection
2. From Inspector, select "From a specific product" and choose a product which have some upsells products.
3. Verify that, as soon as you select a product, correct upsells products are shown on Editor side.
4. Save template
5. Go to frontend and add any products to cart.
6. Checkout and you will be redirected to "Order Confirmation" page.
7. On "Order Confirmation" page, verify that "Upsells" collection always shows upsells products from the product which you selected in Inspector control. It doesn't matter which products you have in Order, it should only show upsells products related to the product which you selected in Inspector control.
8. Perform same testing for "Cross-Sells" collection.

**Case 4: Optional advance testing**
Verify that "Upsells" and "Cross-Sells" collections are working as expected in other contexts i.e.:
- On a new post/page
- On "Single Product" template
- On "Cart" template
- On "Product Catalog" template

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [x] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Product Collection: Add "order" reference for "Upsells" and "Cross-Sells" collections.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

imanish003 committed:
Add order reference to Upsells and Cross-sells

- Added 'order' to usesReference in cross-sells and upsells
- Updated LinkedProductControl for order context
- Enhanced ProductCollection to handle order references

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

imanish003 committed:
Add PHP code for Cross-sells collection

imanish003 committed:
Merge branch 'add/52019-product-collection-add-order-and-cart-reference-to-product-related-collections' of https://github.com/woocommerce/woocommerce into add/52019-product-collection-add-order-and-cart-reference-to-product-related-collections

imanish003 committed:
Fix formatting

imanish003 committed:
Add E2E tests

1. In the Cart Template, it checks that "From products in the cart" is
   selected by default for the "My Custom Collection - Cart Context".

2. In the Order Confirmation Template, it verifies that "From products
   in the order" is chosen by default for the "My Custom Collection -
   Order Context".

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/cross-sells.tsx:
```diff
@@ -21,7 +21,7 @@ const collection = {
 	),
 	keywords: [ 'boost', 'promotion' ],
 	scope: [],
```
I'm fixing the Cross-Sells scope in https://github.com/woocommerce/woocommerce/pull/52209 but In the current shape of this PR, Cross-Sells is unavailable and cannot be tested correctly so I suggest you add this (quality team is testing PRs weekly so there's a chance this PR will be tested without mine and would be perceived as broken):

```suggestion
	scope: [ 'block' ],
```

imanish003 replied:
Good point. I thought about adding it but then I thought your PR might be merged first. But it makes sense to add in this PR as well, in case your PR isn't merged first. Also, It would be easier for reviewer to review this PR.

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -155,13 +159,19 @@ const LinkedProductControl = ( {
 	const isCartLocation = location.type === REFERENCE_TYPE_CART;
 	const hasCartReference = !! usesReference?.includes( REFERENCE_TYPE_CART );
+
+	const REFERENCE_TYPE_ORDER = 'order';
```
Nitpick: seems like `REFERENCE_TYPE_CART` and `REFERENCE_TYPE_ORDER` could be extracted to the top of the file and could be used in `getFromCurrentProductRadioLabel` as well.

imanish003 replied:
Good point! Made the changes in [e65e88a](https://github.com/woocommerce/woocommerce/pull/52258/commits/e65e88a474fb9e72354e9c78f1b30925e36a7adf)

kmanijak commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -2142,9 +2147,14 @@ function ( $collection_args, $query ) {
 					if ( isset( $location['type'] ) && 'product' === $location['type'] ) {
 						$product_reference = array( $location['sourceData']['productId'] );
```
`product_reference` vs `product_references` is back 😆 

Could you either update the var name here to use plural version (the same as in Upsells) or actually make Upsells using singular? 👼 

I know I was strongly opting for plural in Upsells but more than anything I think the var name should be aligned between collections 😅

imanish003 replied:
Haha I will update it to plural i.e. `product_references` in this PR 🤝

kmanijak reviewed (changes requested):
All testing well! However, I had to "enable" Cross-Sells myself by adding `scope: [ 'block' ],`. Please add it so testing steps can be fully tested.

Except for that, I just left some nitpicks. Great work Manish!

imanish003 committed:
Add 'block' scope to cross-sells collection

imanish003 committed:
Extract and use reference type constants

This commit extracts the REFERENCE_TYPE_CART and REFERENCE_TYPE_ORDER
constants to the top of the file and updates their usage throughout
the code, including in the getFromCurrentProductRadioLabel function.

imanish003 committed:
Align variable naming for product references in Product Collection

This commit updates the ProductCollection class to ensure consistency in
variable naming for product references across different collection types,
particularly aligning with the naming convention used in Upsells.

imanish003 commented:
Hey @kmanijak, I have made the changes, and PR is ready for review again.

imanish003 committed:
Merge branch 'trunk' into add/52019-product-collection-add-order-and-cart-reference-to-product-related-collections

kmanijak reviewed (approved):
Thanks for addressing all my comments! <3 Amazing job!

xristos3490 commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -129,10 +133,17 @@ const getFromCurrentProductRadioLabel = (
 		return __( 'From products in the cart', 'woocommerce' );
 	}
 
+	if ( currentLocation === REFERENCE_TYPE_ORDER ) {
```
Is there a reason why a similar `isUsesReferenceIncludesOrder` isn't needed here? Alternatively, is the cart-related one necessary in the first place?

imanish003 replied:
Wow! Great catch. We need it for Order context. I will add it. Thanks for pointing it out 🙇‍♂️

xristos3490 commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -2065,9 +2065,14 @@ function ( $collection_args, $query ) {
 					}
+
+					if ( isset( $location['type'] ) && 'order' === $location['type'] ) {
+						$product_references = $this->get_product_ids_from_order( $location['sourceData']['orderId'] ?? null );
```
We should fallback to `0` here since the `get_product_ids_from_order` expects an `int`

imanish003 replied:
I don't have a strong opinion here, so I added fallback to 0.

xristos3490 commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -2135,19 +2140,24 @@ function ( $acc, $product ) {
+					}
+
+					if ( isset( $location['type'] ) && 'order' === $location['type'] ) {
+						$product_references = $this->get_product_ids_from_order( $location['sourceData']['orderId'] ?? null );
```
Not strictly related to this PR:
Might be best to default to `array()` instead of `null`. Both options could cause the variable to change its type during execution.
```suggestion
						$product_references = $this->get_product_ids_from_order( $location['sourceData']['orderId'] ?? array() );
```
Same applies for L2143.

imanish003 replied:
I am not sure if I understand the issue correctly. I already updated it to fallback to 0 as suggested in [this comment](https://github.com/woocommerce/woocommerce/pull/52258#discussion_r1815151545). Also, this `null` fallback is for argument of `get_product_ids_from_order()` function and not for `product_references` variable. Am I missing something?

xristos3490 replied:
No worries! I noticed a refactor task on the roadmap, so since this is just a minor point, I'll go ahead and resolve the conversation here.

As a general rule, especially after PHP 8, changing a variable's type is considered an antipattern. This will also cause a fatal error if the `strict_types` directive is included. Just keep this in mind!

xristos3490 reviewed (commented):
Nice work, @imanish003!

imanish003 committed:
Add "hadOrderReference" to conditional label display

Key changes:
- Removed `isUsesReferenceIncludesCart` parameter from `getFromCurrentProductRadioLabel`
- Introduced `hasCartReference` and `hasOrderReference` parameters for consistency
- Updated conditional logic to check for both cart and order references
- Adjusted the function call in `LinkedProductControl` to pass both new parameters

imanish003 committed:
Fallback to 0 for orderId in get_product_ids_from_order calls

imanish003 commented:
Hey @xristos3490, I have made the requested changes. Can you please review it again? Please let me know if there is anything that should block this PR from being merged. Thanks again for your detailed feedback 🙇‍♂️

xristos3490 reviewed (approved):
Thanks for tackling my comments, @imanish003! Looks great now!
